databuilder/query_model/transforms.py

```diff
@@ -41,39 +97,27 @@ class PickOneRowPerPatientWithColumns(PickOneRowPerPatient):
 
 
 def apply_transforms(variables):
-    # For algorithmic ease we're going to be mutating the query objects. Since the QM
-    # graph is a notionally immutable structure consisting of "frozen" dataclasses, we need
-    # to do that carefully. In particular:
-    #
-    # 1. We copy the data that is passed in so that callers don't observe side-effects.
-    # 2. We copy the results on the way out to rectify the state of containers that depend
-    #    on object hashes which may have changed unexpectedly (specifically the frozenset used
-    #    to hold selected columns).
-    # 3. When we store QM nodes in equality/hash-sensitive containers during this manipulation
-    #    we use customized versions of those containers which ignore the __eq__() and
-    #    __hash__() implementations provided by dataclasses.
-    variables = copy.deepcopy(variables)
     nodes = all_nodes_from_variables(variables)
     reverse_index = build_reverse_index(nodes)
 
     transforms = [
         (PickOneRowPerPatient, rewrite_sorts),
     ]
 
+    rewriter = QueryGraphRewriter()
     for type_, transform in transforms:
-        apply_transform(type_, transform, nodes, reverse_index)
+        apply_transform(rewriter, type_, transform, nodes, reverse_index)
 
-    variables = copy.deepcopy(variables)  # see comment above
-    return variables
+    return rewriter.rewrite(variables)
 
 
-def apply_transform(type_, transform, nodes, reverse_index):
+def apply_transform(rewriter, type_, transform, nodes, reverse_index):
     for node in nodes:
         if isinstance(node, type_):
-            transform(node, reverse_index)
+            transform(rewriter, node, reverse_index)
 
 
-def rewrite_sorts(node, reverse_index):
+def rewrite_sorts(rewriter, node, reverse_index):
```

Comment:
Do the the transforms need to know about the rewriter?  Or can the transforms return/yield nodes and their replacements?

Or, given there's currently a single transform, is this premature?

Reply:
I think there's a fair bit of rejigging that could be done to the transform code, but I wanted to keep the changes as minimal as possible because I didn't really intend to be working on this at all.